crates/bevy_ecs/src/query/filter.rs

```diff
@@ -376,261 +471,419 @@ macro_rules! impl_query_filter_tuple {
-                        .then(|| {
-                            world.unsafe_world()
-                                 .storages()
-                                 .sparse_sets
-                                 .get(id)
-                                 .debug_checked_unwrap()
-                        }),
-                    last_run,
-                    this_run,
-                }
-            }
+/// A filter on a component that only retains results added after the system last ran.
+///
+/// A common use for this filter is one-time initialization.
+///
+/// To retain all results without filtering but still check whether they were added after the
+/// system last ran, use [`Ref<T>`](crate::change_detection::Ref).
+///
+/// # Examples
+///
+/// ```
+/// # use bevy_ecs::component::Component;
+/// # use bevy_ecs::query::Added;
+/// # use bevy_ecs::system::IntoSystem;
+/// # use bevy_ecs::system::Query;
+/// #
+/// # #[derive(Component, Debug)]
+/// # struct Name {};
+///
+/// fn print_add_name_component(query: Query<&Name, Added<Name>>) {
+///     for name in &query {
+///         println!("Named entity created: {:?}", name)
+///     }
+/// }
+///
+/// # bevy_ecs::system::assert_is_system(print_add_name_component);
+/// ```
+pub struct Added<T>(PhantomData<T>);
 
-            const IS_DENSE: bool = {
-                match T::Storage::STORAGE_TYPE {
-                    StorageType::Table => true,
-                    StorageType::SparseSet => false,
-                }
-            };
+#[doc(hidden)]
+pub struct AddedFetch<'w, T> {
+    table_ticks: Option<ThinSlicePtr<'w, UnsafeCell<Tick>>>,
+    marker: PhantomData<T>,
+    sparse_set: Option<&'w ComponentSparseSet>,
+    last_run: Tick,
+    this_run: Tick,
+}
 
-            const IS_ARCHETYPAL:  bool = false;
+impl<'w, T> Clone for AddedFetch<'w, T> {
+    fn clone(&self) -> Self {
+        Self {
+            table_ticks: self.table_ticks,
+            sparse_set: self.sparse_set,
+            last_run: self.last_run,
+            this_run: self.this_run,
+            marker: PhantomData,
+        }
+    }
+}
 
-            #[inline]
-            unsafe fn set_table<'w>(
-                fetch: &mut Self::Fetch<'w>,
-                &component_id: &ComponentId,
-                table: &'w Table
-            ) {
-                fetch.table_ticks = Some(
-                    $get_slice(
-                        &table
-                            .get_column(component_id)
-                            .debug_checked_unwrap()
-                    ).into(),
-                );
-            }
+/// SAFETY:
+/// `fetch` accesses a single component in a readonly way.
+/// This is sound because `update_component_access` and `update_archetype_component_access` add read access for that component and panic when appropriate.
+/// `update_component_access` adds a `With` filter for a component.
+/// This is sound because `matches_component_set` returns whether the set contains that component.
```

Comment:
Yes, this could probably be more clearly phrased. I'm trying to explain how this type follows the safety rules for filters. 

> /// `fetch` accesses a single component in a readonly way.
This is a statement of a thing that the filter does which is policed by the safety rules
> /// This is sound because `update_component_access` and `update_archetype_component_access` add read access for that component and panic when appropriate.

This is an explanation of how the thing in the first line is okay because the relevant rules for it have been followed. 

The relevant rules are (from world_query.rs)

> For each component readonly accessed by [`fetch`], [`update_component_access`] should add read access unless write access has already been added, in which case it should panic.
> For each component readonly accessed by [`fetch`], [`update_archetype_component_access`] should add read access if that component belongs to the archetype.

The third and fourth lines work in a similar way but are addressing a different set of rules. 

I would love it if someone could help me find a simpler, more readable way of expressing this. Possibly labelling or numbering the rules and then referring to them that way would help?